File: 2021/TypeScript/01.ts
import * as fs from "fs";
import * as path from "path";

function readInput(): string {
  const inputPath = path.join(__dirname, "../input/01.txt");
  return fs.readFileSync(inputPath, "utf8").trim();
}

function part1(input: string): number {
  const numbers = input.split("\r\n").map(Number);

  return numbers.reduce((increase, current, index) => {
    return index > 0 && current > numbers[index - 1] ? increase + 1 : increase;
  }, 0);
}

function part2(input: string): number {
  const numbers = input.split("\r\n").map(Number);
  let measurementIncreases = 0;

  for (let i = 3; i < numbers.length; i++) {
    if (numbers[i] > numbers[i - 3]) {
      measurementIncreases++;
    }
  }

  return measurementIncreases;
}

const input = readInput();

console.log("Part 1:", part1(input));
console.log("Part 2:", part2(input));

File: 2021/Rust/01.rs
mod day_01 {
    use std::fs;
    use std::path::{Path, PathBuf};

    fn read_input() -> String {
        let input_path: PathBuf = Path::new(file!()).parent().unwrap().join("../input/01.txt");
        fs::read_to_string(input_path)
            .expect("Failed to read input file")
            .trim()
            .to_string()
    }

    fn part1(input: &str) -> i32 {
        let numbers: Vec<i32> = input.lines().map(|line| line.parse().unwrap()).collect();
    
        numbers
            .windows(2)
            .filter(|window| window[1] > window[0])
            .count() as i32
    }

    fn part2(input: &str) -> i32 {
        let numbers: Vec<i32> = input.lines().map(|line| line.parse().unwrap()).collect();

        numbers
            .windows(4)
            .filter(|window| {
                let sum1 = window[0] + window[1] + window[2];
                let sum2 = window[1] + window[2] + window[3];
                sum2 > sum1
            })
            .count() as i32
    }

    pub fn run() {
        let input: String = read_input();

        println!("Part 1: {}", part1(&input));
        println!("Part 2: {}", part2(&input));
    }
}

pub use day_01::run;
